pkg/subsystem/linux/maintainers.go

```diff
@@ -0,0 +1,122 @@
+// Copyright 2023 syzkaller project authors. All rights reserved.
+// Use of this source code is governed by Apache 2 LICENSE that can be found in the LICENSE file.
+
+package linux
+
+import (
+	"bufio"
+	"fmt"
+	"io"
+	"strings"
+)
+
+// maintainersRecord represents a single raw record in the MAINTAINERS file.
+type maintainersRecord struct {
+	name            string
+	includePatterns []string
+	excludePatterns []string
+	regexps         []string
+	lists           []string
+	maintainers     []string
+}
+
+func parseLinuxMaintainers(content io.Reader) ([]*maintainersRecord, error) {
+	scanner := bufio.NewScanner(content)
+	// First skip the headers.
+	for scanner.Scan() {
+		line := scanner.Text()
+		if line == "Maintainers List" {
+			// Also skip ------.
+			scanner.Scan()
+			break
+		}
+	}
+	ret := []*maintainersRecord{}
+	for skipComments(scanner) {
+		line := strings.TrimSpace(scanner.Text())
+		record := &maintainersRecord{name: line}
+		ret = append(ret, record)
+		for scanner.Scan() {
+			property := scanner.Text()
+			if property == "" {
+				break
+			}
+			err := applyProperty(property, record)
+			if err != nil {
+				return nil, fmt.Errorf("failed to apply %#v: %w", property, err)
+			}
+		}
+	}
+	if err := scanner.Err(); err != nil {
+		return nil, err
+	}
+	return ret, nil
+}
+
+func skipComments(scanner *bufio.Scanner) bool {
+main:
+	for scanner.Scan() {
+		line := strings.TrimSpace(scanner.Text())
+		if line == "" {
+			continue
+		} else if strings.HasPrefix(line, ".") {
+			for scanner.Scan() {
```

Comment:
Multiple scanner.Scan() loops in different functions, including nested ones, makes it very hard to follow.
Also the current parsing seems to be very sensitive to new lines. E.g. if there is no new line after a comment, we will mis-parse. If there is no new line between subsystems, we will mis-parse.

I think it should be simpler and more robust if you we a single scanner.Scan() and classify lines into subsystem name, property and everything else. Everything else is ignored. Properties are added to the current subsystem. On subsystem name we append the previous one if it got all required properties and start a new subsystem.

Reply:
I've rewritten the parsing code. Is it better now?